src/EventStore.Core/Services/ClusterStorageWriterService.cs

```diff
@@ -131,9 +131,37 @@ public class ClusterStorageWriterService<TStreamId> : StorageWriterService<TStre
 						"ATTEMPT TO TRUNCATE EPOCH WITH COMMITTED RECORDS. THIS MAY BE BAD, BUT IT IS OK IF A NEWLY-ELECTED LEADER FAILS IMMEDIATELY AFTER ELECTION.");
 				}
 
+				Log.Information("Setting the truncate checkpoint to: {truncatePosition} (0x{truncatePosition:X})",
+					message.SubscriptionPosition, message.SubscriptionPosition);
 				Db.Config.TruncateCheckpoint.Write(message.SubscriptionPosition);
 				Db.Config.TruncateCheckpoint.Flush();
 
+				// try to write the new epoch position prior to shutting down to
+				// avoid scanning the transaction log to find a valid epoch
+				// when starting up for truncation
+				var oldEpoch = EpochManager.GetLastEpoch();
+				if (EpochManager.TryTruncateBefore(message.SubscriptionPosition, out var newEpoch)) {
+					Log.Information("Truncated epoch from "
```

Comment:
Not a blocker to me, but if there are no newer epochs on the truncated node, you'll get a log that might be confusing (essentially you're truncating the epoch to the same position as it was before):

```
Truncated epoch from E134@1276418448:{66b39e7c-005b-4c44-be8b-fab3106a839b} to E134@1276418448:{66b39e7c-005b-4c44-be8b-fab3106a839b}
```

Perhaps the log could simply say `Truncated epoch to {newEpochNumber}@{newEpochPosition}:{newEpochId}` and conditionally append `from {oldEpochNumber}@{oldEpochPosition}:{oldEpochId}` if the old epoch was different to the new one? So in a case where the epochs are the same you'll get:

```
Truncated epoch to E134@1276418448:{66b39e7c-005b-4c44-be8b-fab3106a839b}
```

and in cases where they differ it'll be:

```
Truncated epoch to E134@1276418448:{66b39e7c-005b-4c44-be8b-fab3106a839b} from E136@1845951600:{c884fe93-27f7-4f30-9096-5718fff58265}
```